lib/spack/spack/build_environment.py

```diff
@@ -1014,26 +1014,33 @@ def add_modifications_for_dep(dep):
 
 
 def get_cmake_prefix_path(pkg):
-    # Note that unlike modifications_from_dependencies, this does not include
-    # any edits to CMAKE_PREFIX_PATH defined in custom
-    # setup_dependent_build_environment implementations of dependency packages
     build_deps      = set(pkg.spec.dependencies(deptype=('build', 'test')))
     link_deps       = set(pkg.spec.traverse(root=False, deptype=('link')))
     build_link_deps = build_deps | link_deps
     spack_built = []
     externals = []
+
     # modifications_from_dependencies updates CMAKE_PREFIX_PATH by first
     # prepending all externals and then all non-externals
     for dspec in pkg.spec.traverse(root=False, order='post'):
         if dspec in build_link_deps:
             if dspec.external:
-                externals.insert(0, dspec)
+                externals.insert(0, dspec.prefix)
             else:
-                spack_built.insert(0, dspec)
-
-    ordered_build_link_deps = spack_built + externals
-    build_link_prefixes = filter_system_paths(
-        x.prefix for x in ordered_build_link_deps)
+                spack_built.insert(0, dspec.prefix)
+
+    # CMAKE_PREFIX_PATH here cannot come from the outside environment under
+    # normal circumstances, it has to come from spack.  include these second
+    # after the spack_built dependencies but before externals
+    env_mod_list = modifications_from_dependencies(pkg.spec, 'build',
+                                                   custom_mods_only=True)
```

Comment:
This should also set `set_package_py_globals=False`.

With that, and if `modifications_for_dep` skipped `dpkg.setup_dependent_package(spec.package.module, spec)`, this would have no side effects (but I'd need to think about what to suggest to skip the latter - maybe another parameter? That might be confusing for the interface).

Reply:
@scheibelp this will be my first (second? not sure) contribution to spack core. Other than adding the argument you suggested I don't have context to understand what you want me to do here.

Reply:
Sorry yes I need to think on it: my first naive suggestion is that we would need another parameter like `set_package_py_globals` which (when set `=False`) would skip `dpkg.setup_dependent_package(spec.package.module, spec)`, but I'm thinking that having so many parameters to control small parts of the function's behavior will make it confusing to understand, so I'm thinking if I should work on refactoring it (which I could as a commit here if that's alright with you, but I wouldn't have time to do that until next week).